### ISSUE TITLE
JiraIssue and FindingGroup  in models should be decoupled

### ISSUE THREAD

**Issue by jynolen** [DefectDojo/django-DefectDojo#9643]
**Be informative**
I use OpenAPI spec to generate me a nice client without too much effort but the generated models (or the spec in Django) might be wrong

**Bug description**
I use https://openapi-generator.tech/ to generate me a Client  + Models from Swgger Json "/api/v2/oa3/schema/?format=json"
So far all good.

How ever when I try to list my Tests with
`TestsApi(client).tests_list(**kwargs)`

I got the following issue:
```
pydantic_core._pydantic_core.ValidationError: 1 validation error for FindingGroup
jira_issue
  Input should be a valid dictionary or instance of JIRAIssue [type=model_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.6/v/model_type
```

This is because the generated models create something like

```
class FindingGroup(BaseModel):
    """
    FindingGroup
    """ # noqa: E501
    id: StrictInt
    name: Annotated[str, Field(strict=True, max_length=255)]
    test: StrictInt
    jira_issue: JIRAIssue
    __properties: ClassVar[List[str]] = ["id", "name", "test", "jira_issue"]
```

meaning that EVERY finding group MUST have a valid JiraIssue which is not the intended model.
Maybe there is something wrong in the Relationship between JiraIssue and FindingGroup

Workarround (at date)

- Edit the Pydantic model for this

```
class FindingGroup(BaseModel):
    """
    FindingGroup
    """ # noqa: E501
    id: StrictInt
    name: Annotated[str, Field(strict=True, max_length=255)]
    test: StrictInt
    jira_issue: Optional[JIRAIssue]
    __properties: ClassVar[List[str]] = ["id", "name", "test", "jira_issue"]
```

**Steps to reproduce**
1. Retreive shcema.json file from http://defectdojo.localhost/api/v2/oa3/schema/?format=json
2. Run in your favorise docker engine
   a. `podman run --rm -v ${PWD}:/local openapitools/openapi-generator-cli generate  -i /local/schema.json -g python -o /local/out/python`
3. Create some test with finding groups enabled
4. Run Following snippet (should be update to match package name)

```
from defectdojo_ci.openapi import ApiClient, Configuration
from defectdojo_ci.openapi.api import TestsApi

apiToken="xxxxx"
config = Configuration(host=defectdojo.localhost, api_key={"tokenAuth":apiToken}, api_key_prefix={"tokenAuth":"Token"})
client = ApiClient(configuration=config)

print(TestsApi(client ).tests_list(**kwargs).results)
```

**Expected behavior**
Print result

**Deployment method** *(select with an `X`)*
- [ ] Docker Compose
- [x] Kubernetes
- [ ] GoDojo

**Environment information**
 - Operating System: Kubernetes
 - OpenApi Schema from image defectdojo/defectdojo-django:2.31.4
 - 
**Logs** 
No servers logs because it's an issue in the schema models

**Comment by kiblik** [comment]
Thank you @jynolen. Not only for pointing out this problem but also for well-written _Steps to reproduce_ (there had been some imperfections but it was good enough for reproduction).

The problematic part is the following of OAS
```yaml
    FindingGroup:
      type: object
      properties:
        id:
          type: integer
          readOnly: true
        name:
          type: string
          maxLength: 255
        test:
          type: integer
        jira_issue:
          allOf:
            - $ref: '#/components/schemas/JIRAIssue'
          readOnly: true
      required:
        - id
        - jira_issue
        - name
        - test
```
At first, I tried to set `JIRAIssueSerializer` to `required=False`, to remove it from the list of required 
https://github.com/DefectDojo/django-DefectDojo/blob/c5d896b2a606985f41669388962e47409f4866ba/dojo/api_v2/serializers.py#L1396-L1397
This change had no effect on the schema. From some perspective, it makes sense (it is not a request but a response schema) but I had no space to invest time to check and investigate why.

My second try was to set `allow_null=True`. This was the result:
```yaml
    FindingGroup:
      type: object
      properties:
        id:
          type: integer
          readOnly: true
        name:
          type: string
          maxLength: 255
        test:
          type: integer
        jira_issue:
          allOf:
            - $ref: '#/components/schemas/JIRAIssue'
          readOnly: true
          nullable: true
      required:
        - id
        - jira_issue
        - name
        - test
```
`jira_issue` is still in the `required` list but `nullable: true` was added.

When I tried to create python client from original schema, store it (`git init; git add .`) and regenerate with new schema, I have received expected result:
```diff
diff --git a/openapi_client/models/finding_group.py b/openapi_client/models/finding_group.py
index cd9c29f..8c42a4d 100644
--- a/openapi_client/models/finding_group.py
+++ b/openapi_client/models/finding_group.py
@@ -18,7 +18,7 @@ import re  # noqa: F401
 import json
 
 from pydantic import BaseModel, Field, StrictInt
-from typing import Any, ClassVar, Dict, List
+from typing import Any, ClassVar, Dict, List, Optional
 from typing_extensions import Annotated
 from openapi_client.models.jira_issue import JIRAIssue
 from typing import Optional, Set
@@ -31,7 +31,7 @@ class FindingGroup(BaseModel):
     id: StrictInt
     name: Annotated[str, Field(strict=True, max_length=255)]
     test: StrictInt
-    jira_issue: JIRAIssue
+    jira_issue: Optional[JIRAIssue]
     __properties: ClassVar[List[str]] = ["id", "name", "test", "jira_issue"]
 
     model_config = {
@@ -80,6 +80,11 @@ class FindingGroup(BaseModel):
         # override the default output from pydantic by calling `to_dict()` of jira_issue
         if self.jira_issue:
             _dict['jira_issue'] = self.jira_issue.to_dict()
+        # set to None if jira_issue (nullable) is None
+        # and model_fields_set contains the field
+        if self.jira_issue is None and "jira_issue" in self.model_fields_set:
+            _dict['jira_issue'] = None
+
         return _dict
 
     @classmethod
```

I will open PR for this change. But now the question is: Is this the only place with this kind of bug?

**Comment by jynolen** [comment]
Hello,

Thanks for quick update, to be honest at this time I didn't see other model
that triggers the bug.

If others occurs we will be able to reference this issue or your PR .

**Comment by kiblik** [comment]
> Hello,
> 
> Thanks for quick update, to be honest at this time I didn't see other model
> that triggers the bug.
> 
> If others occurs we will be able to reference this issue or your PR .
> 

#9651 was merged and will be released in `2.32.0` (hopefully next week).
I suppose we can close this issue and as @jynolen mentioned, it will happen again, this issue can be referenced and we already know how to fix it.
If somebody is able to do some proactive check (or even write unittests for this), it would be awesome.
